Feature/force single drainage point 190

richpsharp opened the pull request:
This feature fixes #190 by adding a parameter to `fill_pits` called `single_outlet_tuple`. If present this forces the fill to treat that pixel as the only valid drain. Combined with another feature that `max_pixel_fill_count` can be set to -1 to fill non-stop this will force every non-nodata pixel in the raster to drain to the pixel targed in `single_outlet_tuple`. To help identify a candidate I've also added another function `detect_lowest_drain_and_sink` which returns the lowest edge drain and internal sink pixels that could be used in tandem with `single_outlet_tuple`. Here's all the changes:
* added new `single_outlet_tuple` parameter to `fill_pits`
* created new helper function `detect_lowest_drain_and_sink`
* fixed a bug in `fill_pits` that would have filled up the nodata region of a raster in a corner case
* added functionality that if `max_pixel_fill_count` is -1 it will not terminate early and set that to the default parameter for `fill_pits`
* added test cases to cover the above
* updated history describing the above

richpsharp committed:
#190 added feature to drain a single point and a test to cover it and updating history

richpsharp committed:
re #190 upate history

richpsharp committed:
fixes #190 added a detect_lowest_sink_and_drain with test case and history update

richpsharp committed:
#190 making max pixel fill be an optional parameter now

richpsharp committed:
re #190 default that parameter to -1 so we don't get confused

richpsharp committed:
#190 fixed an issue where it was just filling wrong

richpsharp committed:
#190 better coverage on drain test case and fixing bug that would fill nodata

richpsharp committed:
re #190 removing debug write outs and fixing expected result for 5,5

richpsharp committed:
re #190 renaming in the order that the parameters come out and internally changing to sink and drain

richpsharp committed:
re #190 noting that i fixed a bug in fill pits

richpsharp committed:
removing some directories in test cases that were manually generated

richpsharp committed:
merge

richpsharp committed:
Merge branch 'main' into feature/force-single-drainage-point-190

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -696,6 +700,13 @@ def fill_pits(
     cdef int feature_id
 
+    # used to handle the case for single outlet mode
+
```
remove empty line to better relate the comment to the code?

richpsharp replied:
removed in 5aaefd9

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -793,10 +804,10 @@ def fill_pits(
-            yi_root = yi-1+yoff
-            for xi in range(1, win_xsize+1):
-                xi_root = xi-1+xoff
+        for yi in range(win_ysize):
+            yi_root = yi+yoff
+            for xi in range(win_xsize):
+                xi_root = xi+xoff
```
I'm assuming this is where you "fixed a prior bug". Could you just comment why it was iterating from `1` before and why we don't want to do that now? I'd love to know the context for this fix / change.

richpsharp replied:
I just noticed this loop ran from 1 to n+1 but then inside the parameter was always i-1. so just changed it from 0 to n then just use i. There's no functional change here.

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -850,8 +876,12 @@ def fill_pits(
                     yi_q = search_queue.front().yi
                     search_queue.pop()
-                    if search_steps > max_pixel_fill_count:
+                    if (max_pixel_fill_count > 0 and
```
Okay, so if `max_pixel_fill_count == -1`, we don't want to come in here and quit the search early. We are riding it out until we find that single drain point. (just 'typing' out loud)

dcdenu4 replied:
Well I guess if `single_outlet` is `None`, then we're just searching until we find an actual drain, be it nodata, an edge, or an outlet.

dcdenu4 replied:
Or if it's determined it should be filled.

richpsharp replied:
Correct.

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
+        (drain_pixel, drain_height, sink_pixel, sink_height) -
+            two (x, y) tuples with corresponding heights, first
+            list is for edge drains, the second is for pit sinks. The x/y
+            coordinate is in raster coordinate space and `height` is the
```
Should `height` be double ticks here?

richpsharp replied:
It wasn't referring to an actual parameter which is why i didn't have double ticks, just the suffix of those but I changed it in 0fd7e78 from `height` to _height so that's more obvious.

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -611,7 +607,9 @@ def _generate_read_bounds(offset_dict, raster_x_size, raster_y_size):
         dem_raster_path_band, target_filled_dem_raster_path,
-        working_dir=None, long max_pixel_fill_count=_MAX_PIXEL_FILL_COUNT,
+        working_dir=None,
+        long long max_pixel_fill_count=-1,
```
This is probably going to have an impact on InVEST models if the default was `500`. Is there an issue in the InVEST repo mentioning this change and the need to update the models?

richpsharp replied:
Sure, I made an issue here: https://github.com/natcap/invest/issues/560 won't affect any test data and it was a stopgap to only affect runs that were run on degenerate DEMs.

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
+        for yi in range(1, win_ysize+1):
+            yi_root = yi-1+yoff
+            for xi in range(1, win_xsize+1):
+                xi_root = xi-1+xoff
```
Here we are doing the iteration with the 1 offset. Is this still intended given the change above in `fill_dem`?

richpsharp replied:
changed in 40ad0cf

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
+                xi_root = xi-1+xoff
+                # this value is set in case it turns out to be the root of a
+                # pit, we'll start the fill from this pixel in the last phase
+                # of the algorithm
```
Update this comment? I don't think we're doing any filling here.

richpsharp replied:
Removed in 1e3f6bf

dcdenu4 commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
+                f'{current_pixel} of {raster_x_size * raster_y_size} '
+                'pixels complete')
+
+        # search block for locally undrained pixels
```
and sinks

richpsharp replied:
changed in 02f5637

dcdenu4 commented on tests/test_routing.py:
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
+        dem_array = numpy.zeros((11, 11), dtype=numpy.float32)
+        dem_array[0, 0] = -1.0
+        dem_array[1:8, 1:8] = -2.0
+        dem_array[10, 7] = -4
```
should this be `-4.0` for consistency?

richpsharp replied:
changed in c05182d

dcdenu4 commented on tests/test_routing.py:
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
+
+        # outlet tuple at 0,0, just drain one edge
+        expected_array_0_0 = numpy.copy(dem_array)
+        expected_array_0_0[1:8, 1:8] = -1
```
Just my own curiosity here. A plateau of all `-1` will drain to an outlet of also `-1`? I guess I'd expect these to be ever so slightly greater than `-1`.

dcdenu4 replied:
By plateau I guess I mean pit in this instance.

richpsharp replied:
Yes because there's a -1 pixel that pours off the edge, so to drain you only need the other pixels to be at the same height.

dcdenu4 commented on tests/test_routing.py:
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
+        expected_array_0_0[10, 7] = 0
+        expected_array_0_0[8:11, 8:11] = 2.0
+
+        # output tuple at 5,5, it's a massive pit so drain the edges
+        expected_array_5_5 = numpy.copy(dem_array)
+        expected_array_5_5[8:11, 8:11] = 2.0
+        expected_array_5_5[10, 7] = 0
```
We're just testing the algorithm works by draining everything to Sarlacc Pit?

richpsharp replied:
GOOD ONE

dcdenu4 commented on tests/test_routing.py:
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
+        for output_tuple, expected_array in [
+                ((0, 0), expected_array_0_0),
+                ((5, 5), expected_array_5_5),
+                ]:
```
Could we have another test in here or separate that also manipulates the `max_fill_counter` variable using the same fake dem you've used here?

richpsharp replied:
I added two more cases that use a `max_pit_fill_count` of 1 that'll result in no pit filling at all. Added in 75a05f5

dcdenu4 commented on tests/test_routing.py:
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
+            numpy.testing.assert_almost_equal(result_array, expected_array)
+
+    def test_detect_lowest_drain_and_sink(self):
+        """PGP>routing: test detect_lowest_sink_and_drain."""
```
>PGP.routing

richpsharp replied:
fixed in ff64743e1148d989d559e87db6c767465d130753

dcdenu4 reviewed (changes requested):
Hey @richpsharp I just gave this a look through. I followed it pretty well I think and didn't find anything glaring. I have a few comments though that should be minor.

richpsharp committed:
re #190 remove blank line

richpsharp committed:
removing ` and adding _

richpsharp committed:
re #190 change to sinks

richpsharp committed:
re #190 removing extra 1 offset

richpsharp committed:
re #190 changing to 4.0 so it shows its a float

richpsharp committed:
re #190 fixing  .

richpsharp committed:
re #190 getting that fill dist test right

richpsharp committed:
Merge branch 'feature/force-single-drainage-point-190' of github.com:richpsharp/pygeoprocessing into feature/force-single-drainage-point-190

richpsharp committed:
#190 adding a test to ensure that fill dist works on single drain point

richpsharp committed:
re #190 merge history

richpsharp committed:
re #190 removing old ocmment

richpsharp commented:
> Hey @richpsharp I just gave this a look through. I followed it pretty well I think and didn't find anything glaring. I have a few comments though that should be minor.

Thanks @dcdenu4, I think I addressed all your comments. Can you take a second look when you get a chance?

dcdenu4 reviewed (approved):
Thanks Rich! Looks okay to me.
